### PR TITLE
Reduce Docker image size, support passing command line properties

### LINKED ISSUE #33: cannot set projectKey
In the current implementation, I don't see I'm supposed to set the `projectKey` or other variables I might need... ?

### LINKED ISSUE #34: Interface needs to allow passing all properties on docker run command line
The interface to this seems to be very inflexible.  I need to be able to specify on the "docker run" command line all the possible properties that I could pass to sonar-scanner.  It's not useful to make it look for a properties file.  Allow sending all properties on the command line.

### PR DESCRIPTION
**Pull request by konstruktoid**
This PR has similarities with #14 but also aims to reduce the Docker image size, verify the integrity of the downloaded `sonar-scanner-cli-${SONAR_SCANNER_VERSION}-linux.zip`, as well as support command line arguments.

I believe this will close #34 and close #33.

E.g:
```
$ docker run --user="$(id -u):$(id -g)" -it -v "$(PWD):/usr/src" sonarsource/sonar-scanner-cli-alpine -Dsonar.projectKey=XYZ   -Dsonar.sources=.   -Dsonar.host.url=http://host.docker.internal:9000   -Dsonar.login=c0555865f652a94d4d36a85846cf754eddeadbeef
```

```
$ docker images sonarsource/*
REPOSITORY                             TAG                 IMAGE ID            CREATED             SIZE
sonarsource/sonar-scanner-cli-alpine   latest              6571ed849525        10 minutes ago      258MB
sonarsource/sonar-scanner-cli          latest              93d07d3bbc89        3 months ago        406MB
```

### CODE REVIEW

**Comment by konstruktoid** [comment]
Probably closes #26 as well.

File: README.md [hunk 1]
```diff
@@ -21,71 +21,63 @@ Licensed under the [GNU Lesser General Public License, Version 3.0](http://www.g
...
-
-Caching is actually shared between projects when running the scanner natively as the `.sonar` is created in the home directory of the current user (eg. `/home/my_user/.sonar`).
-
-You can reproduce this behavior, by adding the following to the `docker run` command:
+When running the scanner with this image, this `.sonar` directory is created in the project's directory. This implies caching is not happening accross analysis of multiple projects.
 
-```
--e SONAR_USER_HOME=/usr/.sonar -v "/home/my_user/.sonar:/usr/.sonar"
-```
+Caching is actually shared between projects when running the scanner natively as the `.sonar` is created in the home directory of the current user.
```
Comment (lenntt) [inline README.md]: This piece of text now confuses me a bit more, is `.sonar` created in the project directory or in the home directory? 
How do I configure it? 
Linking to #27, as this isn't per se part of this PR.

Comment (konstruktoid) [reply README.md]: Hi @lenntt, 
`.sonar` is created in the project directory when using the Docker container, so yes, that needs to be rewritten.

I don't really know if it is possible to configure.

I suggest that you add the following to any `.gitignore` files.

```
.sonar
.scannerwork
```

**Comment by glimchb** [comment]
Any chance this is going to be merged soon? 
what is the hold up from @michal-duda-sonarsource  @simonbrandhof  @jacek-poreda-sonarsource @amelie-foulc-sonarsource ?

File: README.md (hunk 1 above)
Comment (konstruktoid) [reply README.md]: @lenntt updated the README.

File: README.md [hunk 2]
```diff
@@ -2,91 +2,131 @@
...
 
-If you can't mount the project into this directory (eg. in Gitlab CI), use `SONAR_PROJECT_BASE_DIR` to specify another location in the container.
+### Caching and scanner user home directory
 
-```
--e SONAR_PROJECT_BASE_DIR=/build
-```
+When running the scanner with this image, the `.sonar` directory is created in
+the project's directory. This implies caching is not happening accross analysis
+of multiple projects.
```
Comment (lenntt) [inline README.md]: I believe this isn't true anymore - `SONAR_USER_HOME` isn't set anywhere anymore (used to be in `entrypoint.sh`), so I believe the default behavior is using .sonar/ in USER_HOME directory?

Comment (konstruktoid) [reply README.md]: Wall of terminal output:
```
$ ls -1
Dockerfile
LICENSE
README.md
__pycache__
cacheview
images
renovate.json
requirements.txt
tests
tox.ini
$ docker run --user="$(id -u):$(id -g)" -it -v "$(PWD):/usr/src" sonarsource/sonar-scanner-cli -Dsonar.projectKey=ansible_cacheview  -Dsonar.sources=.  -Dsonar.host.url=http://host.docker.internal:9000  -Dsonar.login=52a56ed0d19c6283413ee1c07933c160c9f1da86
INFO: Scanner configuration file: /opt/sonar-scanner/conf/sonar-scanner.properties
INFO: Project root configuration file: NONE
INFO: SonarQube Scanner 4.2.0.1873
INFO: Java 11.0.5 Alpine (64-bit)
INFO: Linux 4.19.76-linuxkit amd64
**INFO: User cache: ?/.sonar/cache**
INFO: SonarQube server 7.9.2
INFO: Default locale: "en_US", source code encoding: "UTF-8" (analysis is platform dependent)
INFO: Load global settings
INFO: Load global settings (done) | time=149ms
INFO: Server id: BF41A1F2-AXAfx-3rzRBL7oqZsEle
**INFO: User cache: /usr/src/?/.sonar/cache**
INFO: Load/download plugins
INFO: Load plugins index
INFO: Load plugins index (done) | time=93ms
INFO: Load/download plugins (done) | time=18655ms
INFO: Process project properties
INFO: Execute project builders
INFO: Execute project builders (done) | time=7ms
INFO: Project key: ansible_cacheview
INFO: Base dir: /usr/src
**INFO: Working dir: /usr/src/.scannerwork**
INFO: Load project settings for component key: 'ansible_cacheview'
[... OUTPUT REMOVED ...]
$ ls -1
'?'
Dockerfile
LICENSE
README.md
__pycache__
cacheview
images
renovate.json
requirements.txt
tests
tox.ini
$ ls \?/.sonar/
_tmp  cache
$ ls \?/.sonar/cache/
0c3ec093156f0350edc21e2a1d7ef2ee  489f1b0ff745f9a744c23a97445caf98  59d98c94277e5faa8377ba521e440eba  84bac7bdcd3d78a2cd7037bfd70fcc3c	b8b93b70f10a74c6b3ab95ef8691c065
2d1a20d5ed66aa7f1273836a29fa08d4  4a3cc03335859f0839de090b0ed35956  6ead78e9824cd63c2952b34ee1f1231c  866bb1adbf016ea515620f1aaa15ec53	cb2c8bc61a566b6f9bd7f4a8d9518a25
2d9994a460180757f3fbad54f03e818e  511f59a24e74f40d7e810467f1e3a253  716a52eef1e2edefeafa4c83c9c33a22  90da8f56fbbbb3624b8a3f6f1075860a	e1977f8aaefcb2066437d32de0b347b0
37ae8929672c56a67919c60df53b4283  513697d7dc792f02b10b1f4bcb9b48ab  73d90ce62c95649d02af4292588b027d  91f25bbf08401b93d651d5f3653d2e0e
45e3bfcd65e7578ed4cd89604de3b06f  5246ba7166e8e9ab17d08a6514e9a086  74dfad9fb80c653ed5c959c3b2ed5a4a  _tmp
```

Comment (konstruktoid) [reply README.md]: `-Dsonar.projectBaseDir=/usr/src` is set as the Docker `CMD`, and the project directory is mounted into that directory in which the cache is written.

Comment (lenntt) [reply README.md]: oh, hah, I was wrong. `?/.sonar` 🤣!

**Comment by konstruktoid** [comment]
How this PR is related to https://jira.sonarsource.com/browse/MMF-1931, @pierre-guillot-sonarsource (https://github.com/SonarSource/sonar-scanner-cli-docker/issues/41#issuecomment-600499418)

_"Use the smallest base image as possible"_, maybe not the smallest, but commonly used and smaller than the absolute majority: https://github.com/SonarSource/sonar-scanner-cli-docker/blob/28ad6a9af1ec25c37f5e79b8d54d7edd255c49fa/4/Dockerfile#L1, https://github.com/konstruktoid/Docker/blob/master/Scripts/docker_image_result#L13-L17

_"Do not package non-required tools inside the Docker image"_, all non-required applications are removed after usage: https://github.com/SonarSource/sonar-scanner-cli-docker/blob/28ad6a9af1ec25c37f5e79b8d54d7edd255c49fa/4/Dockerfile#L37-L39

_"Download packages securely using GPG and certificates"_, sonar scanner .zip filed is verified: https://github.com/SonarSource/sonar-scanner-cli-docker/blob/28ad6a9af1ec25c37f5e79b8d54d7edd255c49fa/4/Dockerfile#L27-L29

_"Use COPY but not ADD"_, `ADD` is used instead of `COPY` since we don't want to include the .zip files in the repo due to the size: https://github.com/SonarSource/sonar-scanner-cli-docker/blob/28ad6a9af1ec25c37f5e79b8d54d7edd255c49fa/4/Dockerfile#L27-L29

_"Apply security best-practices", all good guidelines (but I'm a bit biased here :)) but those are most run time issues, the `docker run` instructions should be updated with `--cap-drop` options etc.

**Comment by pierre-guillot-sonarsource** [comment]
ping @michal-duda-sonarsource :)

**Comment by kav** [comment]
Probably also worth mentioning; these things improve iteratively. Rejecting an improvement when there is an internal large batch effort to fix everything all at once only alienates the community. Even if you merge everyone's changes and then turn around and improve them further you'll build goodwill and support. I'd be happy to chat about ways to foster a healthy community if you'd like to talk through this a bit more.

Also as a complete aside; the distroless stuff looked really promising in my own work as well but it feels a bit under resourced and not as well maintained as I'd like. YMMV

**Comment by michal-duda-sonarsource** [comment]
First of all: thank you for the contribution. This indeed looks like a very nice PR but it's very drastic in the number of changes it brings. Something like this needs validation on our side to ensure it works well with our analysers for which we currently have no bandwidth. It also breaks backwards compatibility with the current image so at the very least the upgrade process needs to be documented. It would be much better if this PR only added the support of passing command line parameters to existing image.

We are thinking about switching our base image and `alpine` is an option but we want to evaluate other images as well (like `distroless`) before making the switch.

**Comment by petemounce** [comment]
@michal-duda-sonarsource in its current state, this container is hard to use within CI, where there are secret properties in play.

For example
* Supplying a token secret
  * If this happens via `SONAR_TOKEN` it's easy to steal via `env | sort | grep "TOKEN"` etc.
    * ... but the container doesn't support `SONAR_TOKEN`, only the username/password combo which if leaked can allow more access than I might want, I think.
  * If this happens via being present inside a `sonar-project.properties` file in source control, it becomes un-secret
  * If this happens via `-Dsonar.token` on the command line, it becomes un-secret by being encoded into a shell script in source control
    * unless I use something else to fetch it to a temporary file with an unguessable name, then `-D sonar.token=$(cat $my_file)` with a `trap 'rm -rf $my_file' EXIT` prior
* Supplying dynamic properties for PR analysis

Since the container is in beta, I'd be in favour of breaking backwards compatibility and pushing this PR forward.

**Comment by michal-duda-sonarsource** [comment]
@petemounce Just to clairfy: SONAR_TOKEN is supported by current image. Or rather it's supported natively by SQ, nothing had to be done on the docker side for it to work.

**Comment by konstruktoid** [comment]
Closing due to https://github.com/SonarSource/docker-sonarqube/pull/379#issuecomment-604570311 and https://github.com/SonarSource/docker-sonarqube/blob/master/README.md#contributing.

**Pull request closed without merging**